rfcs/0002-mvp.md

```diff
@@ -0,0 +1,159 @@
+- `.supabase/config.txt` - CLI config and details
+- `.supabase/docker-compose.yml` - for running Supabase locally. We may not even need this, as we can just run it directly from the `npm_modules` folder, which would prevent conflicts when we add upgrades or make changes. If we do generate a file, this should be stored in the `.supabase` folder so that it doesn't conflict with the user's docker-compose.yml file. When the user runs `supa eject`, it can move the file out to their root folder (similar to a `cra eject` model), and then the user can manage this themselves.
+
+
+For the database scaffold, [this](https://gitlab.com/delibrium/delibrium-postgrest/-/tree/master/database) format is very nice, although I think we can make it a bit clearer. Here is a suggested format:
+
+```sh
+│   # Data: database fixed/seed data.
+├── data
+│   ├── some_data.{sql|csv|json}        
+│   # Models: Tables and Views
+├── models
+│   ├── public.table                    # Personal preference to explicitly add the schema, and we should prefer dot notation over nesting.
+│   │   ├── data.{sql|csv|json}         # For convenience, table specific data can be stored here too.   
+│   │   ├── model.sql                   # This is the core. We should also figure out a nice, readable format for SQL files. Should include constraints and indexes and table/column comments.
+│   │   ├── permissions.sql             # Grants and Policies.
+│   │   ├── test.sql                    # PGTAP tests?
+│   │   └── triggers.sql                # table triggers 
+│   # Functions/RPC: functions. At the moment the function are written in SQL. If we want to support something like plv8, we would have to include a builder (like webpack) to bundle the function.
+├── functions                                
+│   ├── public.func1.sql                
+│   ├── public.func2.sql│   
+│   # Postgres roles and grants.
+├── roles                                
+│   ├── role_name.sql    
+│   # Types: custom database types
+└── types
+    ├── _init.sql
+    ├── public.some_type.sql
+    └── extensions.some_type.sql
+```
+
+
+### Branching
+
+This is going to be a key part of Supabase development. Since so many of our users are on Vercel/Netlify, we should have the ability to support [branch/preview-development](https://vercel.com/docs/platform/deployments).
+
+We can do this by connecting to a different database, an idea that I got from [pgsh](https://github.com/sastraxi/pgsh). For example:
+
+- `postgresql://postgres@localhost:5432/master`
+- `postgresql://postgres@localhost:5432/develop`
+- `postgresql://postgres@localhost:5432/chore_add_types`
+
+This is the same from CI (like github actions), except that instead of running against a local database, it is running against the supabase database. We will need a way inside our config to match branch names to database names too. For example 
+
+
+### Dump
+
+```sh
+supa db dump                # dumps the currently active (local) database to disk
```

Comment:
I used `postgres-meta` to create a small mock here: https://github.com/supabase/cli/pull/3

I'm not entirely sure if it's the right approach. On the one hand, we have full control to manage the layout/structure of the dump. On the other hand, we will need to cover all the nuances of Postgres. 

Perhaps it's better rely on `pgdump` from the outset. doing some sort of system where we either:

Option 1:

1. Run pg_dump on the entire database
2. Run some sort of AST parser on the full dump
3. Clean it up and process it into separate files

or Option 2:

1. Run `pg_dump` on individual tables/functions/types etc (eg: `pg_dump -Fp -t $table -f $table.dump;`). We can use `postgres-meta` to fetch the "structure" of the database
2. Run some some sort of "prettier" formatter to make it look readable


I'm leaning towards option 2, as it also give the develop better control. For example they might want to dump just their "functions".

Reply:
> For example they might want to dump just their "functions".

AFAICT, `pg_dump` cannot dump only functions([SO ref](https://stackoverflow.com/questions/13758003/how-to-take-backup-of-functions-only-in-postgres)).

So if we go the above way, we'd need a combination of 1 and 2

---

Maybe there's another way.

If you look at the Schema Diff output image in https://github.com/supabase/cli/pull/2#discussion_r591909137, you'll notice that it also gives the source DDL. And all the database objects are neatly separated.

The Flask endpoint of the pgadmin schema diff process gives this [big json](https://gist.github.com/steve-chavez/e0f857cb223cfc8a539cd1feceb0f102) as its response. The diff related to the above image is located [here](https://gist.github.com/steve-chavez/e0f857cb223cfc8a539cd1feceb0f102#file-schema_diff_output-json-L4953-L4967).

Notice that the ddl output also has COMMENTs included and it's pretty printed(has newlines).

Checking the big json, all the different types of db objects(event triggers, tables, functions, etc) are there. If we could (ab)use that output, we'd have most of what we need already organized.